migrator/src/main/scala/com/scylladb/migrator/alternator/DynamoDBInputFormat.scala

```diff
@@ -0,0 +1,86 @@
+package com.scylladb.migrator.alternator
+import org.apache.commons.logging.LogFactory
+import org.apache.hadoop.dynamodb.DynamoDBConstants
+import org.apache.hadoop.mapred.{ InputSplit, JobConf }
+
+/**
+  * Specializes the split strategy:
+  *    - do not bound the maximum number of partitions by the available memory per node
+  *    - use as many partitions as the number of scan segments
+  *    - by default, create segments that split the data into 128 MB chunks
+  */
+class DynamoDBInputFormat extends org.apache.hadoop.dynamodb.read.DynamoDBInputFormat {
+
+  private val log = LogFactory.getLog(classOf[DynamoDBInputFormat])
+
+  override def getSplits(conf: JobConf, desiredSplits: Int): Array[InputSplit] = {
+    val readPercentage = conf.getDouble(
+      DynamoDBConstants.THROUGHPUT_READ_PERCENT,
+      DynamoDBConstants.DEFAULT_THROUGHPUT_PERCENTAGE.toDouble)
+    if (readPercentage <= 0) {
+      sys.error(s"Invalid read percentage: ${readPercentage}")
+    }
+    log.info(s"Read percentage: ${readPercentage}")
+    val maxReadThroughputAllocated = conf.getInt(DynamoDBConstants.READ_THROUGHPUT, 1)
+    val maxWriteThroughputAllocated = conf.getInt(DynamoDBConstants.WRITE_THROUGHPUT, 1)
+    if (maxReadThroughputAllocated < 1.0) {
+      sys.error(
+        s"Read throughput should not be less than 1. Read throughput percent: ${maxReadThroughputAllocated}")
+    }
+
+    val configuredReadThroughput =
+      math.max(math.floor(maxReadThroughputAllocated * readPercentage).intValue(), 1)
+
+    val tableSizeBytes = conf.getLong(DynamoDBConstants.TABLE_SIZE_BYTES, 1)
+    val numSegments =
+      getNumSegments(maxReadThroughputAllocated, maxWriteThroughputAllocated, tableSizeBytes, conf)
+
+    val numMappers = getNumMappers(numSegments, configuredReadThroughput, conf)
+
+    log.info(s"Using ${numSegments} segments across ${numMappers} mappers")
+
+    getSplitGenerator().generateSplits(numMappers, numSegments, conf)
+  }
+
+  override def getNumSegments(tableNormalizedReadThroughput: Int,
+                              tableNormalizedWriteThroughput: Int,
+                              currentTableSizeBytes: Long,
+                              conf: JobConf): Int = {
+    // Use configured scan segment if provided
+    val configuredScanSegment = conf.getInt(DynamoDBConstants.SCAN_SEGMENTS, -1)
+    if (configuredScanSegment > 0) {
+      val numSegments =
+        math.max(
+          math.min(configuredScanSegment, DynamoDBConstants.MAX_SCAN_SEGMENTS),
+          DynamoDBConstants.MIN_SCAN_SEGMENTS
+        )
+      log.info(
+        s"Using number of segments configured using ${DynamoDBConstants.SCAN_SEGMENTS}: ${numSegments}")
+      numSegments
+    } else {
+      // split into segments of at most 100 MB each (note: upstream implementation splits into 1 GB segments)
```

Comment:
The upstream implementation uses segments of 1 GB each.

In my tests the throughput is the highest with `n` segments of about 100 MB where `n` is also divisible by the total number of CPUs of the Spark cluster (e.g., in my case I had 8 cores, and the best throughput was observed with 16 segments of 106 MB to transfer 1.7 GB of data in total).

Reply:
if you use faster cpu, memory and network
100MB might be bigger

also for big DBs you might end up with lots of tasks that will expect more memory on driver and master

Reply:
can we make this configurable? @julienrf ?

Reply:
Yes, good idea!

Reply:
btw. audience project had similar calculation, 
see
https://github.com/audienceproject/spark-dynamodb/blob/816c6e6d8a250a2d8b700761b94a198a872a7ea6/src/main/scala/com/audienceproject/spark/dynamodb/connector/TableConnector.scala#L64

seems they used ~128MB

Reply:
okay, you can always use numSegments  to override manually ...

Reply:
> okay, you can always use numSegments to override manually ...

Yes, you can already manually set `scanSegments` and the logic below is used only in case it is missing. We could also provide a `segmentSize` setting which could be an alternative to `scanSegments` to control the partitioning. The benefits would be that the same configuration (e.g. `segmentSize: 100`) could be used with different tables and it would automatically compute an appropriate number of segments that fit the desired size. Whereas with `scanSegments` you have to adapt its value to every table you migrate (because you want to adjust the number of segments according to each table size).

Reply:
I’ve created #146 to track this possible feature that we could implement in the future.